### PR TITLE
kubernetes components using leader election to leases

### PR DESCRIPTION
**Pull request by aimuz**
#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

xref #80289

#### Does this PR introduce a user-facing change?

```release-note
Kubernetes components that perform leader election now only support using Leases for this.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-base/config/validation/validation.go [hunk 1]
```diff
@@ -48,9 +48,6 @@ func ValidateLeaderElectionConfiguration(cc *config.LeaderElectionConfiguration,
 	if cc.LeaseDuration.Duration <= cc.RenewDeadline.Duration {
 		allErrs = append(allErrs, field.Invalid(fldPath.Child("leaseDuration"), cc.RenewDeadline, "LeaseDuration must be greater than RenewDeadline"))
 	}
-	if len(cc.ResourceLock) == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/config/validation/validation.go:51]: restore this check, this is generic validation multiple components (even out of tree components) can use

File: staging/src/k8s.io/component-base/config/options/leaderelectionconfig.go [hunk 2]
```diff
@@ -40,10 +40,9 @@ func BindLeaderElectionFlags(l *config.LeaderElectionConfiguration, fs *pflag.Fl
 	fs.DurationVar(&l.RetryPeriod.Duration, "leader-elect-retry-period", l.RetryPeriod.Duration, ""+
 		"The duration the clients should wait between attempting acquisition and renewal "+
 		"of a leadership. This is only applicable if leader election is enabled.")
-	fs.StringVar(&l.ResourceLock, "leader-elect-resource-lock", l.ResourceLock, ""+
-		"The type of resource object that is used for locking during "+
-		"leader election. Supported options are 'leases', 'endpointsleases' "+
-		"and 'configmapsleases'.")
+	fs.StringVar(&l.ResourceLock, "leader-elect-resource-lock", "leases", ""+
+		"Deprecated: The type of resource object that is used for locking during "+
+		"leader election. This option setting will not take effect, Options locked to 'leases'")
```
Comment (liggitt) [inline staging/src/k8s.io/component-base/config/options/leaderelectionconfig.go:45]: revert this change or make it conditional on a parameter that kube-scheduler/kube-controller-manager, etc can pass in. We will continue supporting endpointsleases and configmapsleases for out of tree callers.

File: cmd/kube-scheduler/app/options/options.go [hunk 3]
```diff
@@ -163,9 +163,6 @@ func (o *Options) ApplyLeaderElectionTo(cfg *kubeschedulerconfig.KubeSchedulerCo
 	if leaderelection.Changed("leader-elect-retry-period") {
 		cfg.LeaderElection.RetryPeriod = o.LeaderElection.RetryPeriod
 	}
-	if leaderelection.Changed("leader-elect-resource-lock") {
-		cfg.LeaderElection.ResourceLock = o.LeaderElection.ResourceLock
-	}
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options.go:168]: revert this change, we still want to use the config field

instead, validate in ValidateKubeSchedulerConfiguration that the field is set to `leases` and fail if it isn't (so a config file using some other lease resource type fails instead of silently ignoring the now-unsupported lease type)

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 4]
```diff
@@ -277,7 +277,6 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
 
 	// Start the main lock
 	go leaderElectAndRun(c, id, electionChecker,
-		c.ComponentConfig.Generic.LeaderElection.ResourceLock,
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/controllermanager.go:280]: keep using/plumbing `LeaderElection.ResourceLock`, and add validation in GenericControllerManagerConfigurationOptions#Validate that if LeaderElection.ResourceLock is set, it is set to `leases`

**Comment by fedebongio** [comment]
/triage accepted
thanks @liggitt !

**Comment by sftim** [comment]
```markdown
Kubernetes components that perform leader election now only support using Leases for this.
```

And for extra docs:
```markdown
[Leases](https://kubernetes.io/docs/concepts/architecture/leases/) in Kubernetes
```

**Comment by aimuz** [comment]
/release-note-edit 

```release-note
Kubernetes components that perform leader election now only support using Leases for this.
```
